### PR TITLE
Feature (SDKS-288): allow configurable server paths

### PR DESCRIPTION
**Pull request by cerebrl**
NOTE: This PR piggybacks off of the Device Profile PR as I needed the more granular E2E testing from that work. Please ignore the additional `fr-device` code related changes.

Summary:

This feature provides the capability of configuring custom paths for SDK related endpoints: authentication, authorization and more. It is based off of this comment from a user: https://github.com/ForgeRock/forgerock-javascript-sdk/commit/1d6cc6d668f5379c358a91838a1e5e248c0b0c8b?email_token=AATELJMDFCIFN5UHAKFH2N3RFUMSNA5CNFSM4LAMSRZKYY3PNVWWK3TUL52HS4DFVVBW63LNNF2EG33NNVSW45FKMNXW23LFNZ2F62LEZYBD3ZDN#r37610605.

Implementation:

This adds a utility function for building the correct path for an endpoint that is used throughout the codebase for uniformity.

TODOs:

- [x] Remove comments that bookmark paths to edit

### CODE REVIEW

File: src/util/url.ts [hunk 1]
```diff
@@ -13,6 +15,28 @@ function getBaseUrl(url: URL): string {
...
+    revoke: `oauth2/${realmUrlPath}/token/revoke`,
+    sessions: `json/${realmUrlPath}/sessions/`,
+  };
+  if (customPaths && customPaths[endpoint]) {
+    return customPaths[endpoint];
+  } else {
+    return defaultPaths[endpoint];
+  }
+}
+
```
Comment (cerebrl) [inline src/util/url.ts:39]: This is the core of the customizable server path feature. All code paths that result in building a URL for fetch goes through this utility function.

File: tests/e2e/app/custom-paths/autoscript.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      paths: {
+        authenticate: 'auth/authenticate',
+        authorize: 'auth/authorize',
+        accessToken: 'auth/accessToken',
+        endSession: 'auth/endSession',
+        userInfo: 'auth/userInfo',
+        revoke: 'auth/revoke',
+        sessions: 'auth/sessions',
+      },
+    },
```
Comment (cerebrl) [inline tests/e2e/app/custom-paths/autoscript.js:35]: Here's the implementation of the server path configuration.

File: tests/e2e/app/custom-paths/autoscript.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  });
+
+  console.log('Initiate first step with `undefined`');
+  rxjs
+    .from(forgerock.FRAuth.next())
```
Comment (cerebrl) [inline tests/e2e/app/custom-paths/autoscript.js:40]: Custom paths is tested with a full OAuth flow.

File: src/http-client/util.ts [hunk 4]
```diff
@@ -54,7 +60,8 @@ function getTxnIdFromURL(urlString: string): string {
   const url = new URL(urlString);
   const value = url.searchParams.get('authIndexValue') || '';
   const parser = new DOMParser();
-  const doc = parser.parseFromString(value, 'text/xml');
+  const decodedValue = decodeURIComponent(value);
```
Comment (cerebrl) [inline src/http-client/util.ts:63]: Found a bug in the (currently unsupported) decoding of IG's XML response by accident :)

File: src/util/url.ts [hunk 5]
```diff
@@ -13,6 +15,32 @@ function getBaseUrl(url: URL): string {
...
+    // TypeScript is not correctly reading the condition above
+    // It's thinking that customPaths[endpoint] may result in undefined
+    // eslint-disable-next-line
+    // @ts-ignore
+    return customPaths[endpoint];
+  } else {
+    return defaultPaths[endpoint];
+  }
+}
+
```
Comment (cerebrl) [inline src/util/url.ts:43]: Core of configurable path code.

File: src/config/interfaces.ts [hunk 6]
```diff
@@ -21,11 +13,42 @@ interface ConfigOptions {
   tree?: string;
 }
 
+type ConfigurablePaths =
```
Comment (adam-cyclones) [inline src/config/interfaces.ts:16]: Would this work better?
``` Typescript
interface CustomPathConfig {
  authenticate?: string;
  authorize?: string;
  accessToken?: string;
  endSession?: string;
  userInfo?: string;
  revoke?: string;
  sessions?: string;
}

type ConfigurablePaths = keyof CustomPathConfig;

const foo: ConfigurablePaths = 'accessToken' ✅ 
```

File: src/util/url.ts [hunk 7]
```diff
@@ -13,6 +15,32 @@ function getBaseUrl(url: URL): string {
   return baseUrl;
 }
 
+function getEndpointPath(
+  endpoint: ConfigurablePaths,
+  realmPath?: string,
+  customPaths?: CustomPathConfig,
+): string {
+  const realmUrlPath = getRealmUrlPath(realmPath);
+  const defaultPaths = {
```
Comment (adam-cyclones) [inline src/util/url.ts:24]: There an interface for this at `src/config/interfaces.ts`

File: src/util/url.ts [hunk 8]
```diff
@@ -13,6 +15,32 @@ function getBaseUrl(url: URL): string {
...
+    userInfo: `oauth2/${realmUrlPath}/userinfo`,
+    revoke: `oauth2/${realmUrlPath}/token/revoke`,
+    sessions: `json/${realmUrlPath}/sessions/`,
+  };
+  if (customPaths && customPaths[endpoint]) {
+    // TypeScript is not correctly reading the condition above
+    // It's thinking that customPaths[endpoint] may result in undefined
+    // eslint-disable-next-line
+    // @ts-ignore
+    return customPaths[endpoint];
```
Comment (adam-cyclones) [inline src/util/url.ts:38]: return (customPaths as CustomPathConfig)[endpoint]

might be cleaner?

File: tests/e2e/app/basic-login/autoscript.js [hunk 9]
```diff
@@ -39,13 +39,27 @@
...
         (step) => step,
       ),
       rxjs.operators.delay(delay),
+      rxMergeMap((step) => {
+        return forgerock.SessionManager.logout();
+      }),
+      rxMap((response) => {
+        if (response.ok) {
+          console.log('Logout successful.');
+          document.body.innerHTML = `<p class="Logged_Out">Logout successful</p>`;
```
Comment (adam-cyclones) [inline tests/e2e/app/basic-login/autoscript.js:58]: A regular string would be better due to lack of interpolation.

File: src/config/interfaces.ts (hunk 6 above)
Comment (cerebrl) [reply src/config/interfaces.ts:16]: Very nice! You just taught me something about TypeScript I didn't know about. Thanks!

File: src/util/url.ts (hunk 7 above)
Comment (cerebrl) [reply src/util/url.ts:24]: Being that this is an object literal, I consider it to have its own interface. I'm a big fan of type inference for literally defined objects like this.

File: src/util/url.ts (hunk 8 above)
Comment (cerebrl) [reply src/util/url.ts:38]: I tried this and a few other things, and unfortunately, it just doesn't like this block of code. It still refuses to properly understand the conditional, so I still need the comments above.

File: tests/e2e/app/basic-login/autoscript.js (hunk 9 above)
Comment (cerebrl) [reply tests/e2e/app/basic-login/autoscript.js:58]: 👍

**Pull request merged**